File: ruby/palindrome-products/palindrome_products.rb
class Palindromes

  def initialize(factor_range)
    @min_factor = factor_range[:min_factor] || 1
    @max_factor = factor_range[:max_factor]
  end

  def generate
    return self unless @palindrome_products.nil?
    @palindrome_products = []
    (@min_factor..@max_factor).to_a.repeated_permutation(2).each do |a, b|
      if is_palindrome(a*b)
        existing_palindrome = @palindrome_products.find { |e| e.value == a*b } rescue nil
        existing_palindrome.nil? ? @palindrome_products << PalindromeProduct.new(a*b, [[a,b]]) : existing_palindrome.factors << [a,b]
      end
    end
    return self
  end

  def largest
    @palindrome_products.find { |e| e.value == @palindrome_products.map(&:value).max }
  end

  def smallest
    @palindrome_products.find { |e| e.value == @palindrome_products.map(&:value).min }
  end

  class PalindromeProduct
    attr_reader :value, :factors
    def initialize(value, factors)
      @value = value
      @factors = factors
    end
  end

  private
  def is_palindrome(num)
    num.to_s.reverse.to_i == num
  end
end

File: python/palindrome-products/palindrome_products.py
def largest(min_factor, max_factor):
    check_parameters(min_factor, max_factor)
    products = {}
    for i in range(max_factor, min_factor -1, -1):
        for j in range(i, min_factor - 1, -1):
            k = i * j
            if len(products.keys()) > 0 and k < list(products.keys())[0]:
                break
            if list(str(k)) == list(reversed(str(k))):
                if k in products:
                    products[k].append([i, j])
                else:
                    products[k] = [[i, j]]
    
    product_factors = sorted([(k, v) for k, v in products.items()])
    return product_factors[-1] if len(product_factors) > 0 else [None, []]

def smallest(min_factor, max_factor):
    check_parameters(min_factor, max_factor)
    products = {}
    for i in range(min_factor, max_factor + 1):
        for j in range(i, max_factor + 1):
            k = i * j
            if len(products.keys()) > 0 and k > list(products.keys())[0]:
                break
            if list(str(k)) == list(reversed(str(k))):
                if k in products:
                    products[k].append([i, j])
                else:
                    products[k] = [[i, j]]
    
    product_factors = [(k, v) for k, v in products.items()]
    return product_factors[0] if len(product_factors) > 0 else [None, []]

def check_parameters(min_factor, max_factor):
    if min_factor > max_factor:
        raise ValueError('invalid factors')